### PR TITLE
ddl: Record the start key of the batch job in delete keys

### PR DESCRIPTION
**Pull request by zimulala**
test on TiKV (start 3 TiKV)
delete 200000 rows before vs after:
old TiDB server takes 0.11-0.21s  per 1024 rows
current TiDB server takes 0.02-0.04s   per 1024 rows

test on goleveldb
delete 204800 rows, old server takes 100s and new server takes 11s.

### CODE REVIEW

File: ddl/schema.go [hunk 1]
```diff
@@ -135,14 +137,14 @@ func getIDs(tables []*model.TableInfo) []int64 {
 }
 
 func (d *ddl) delReorgSchema(t *meta.Meta, job *model.Job) error {
+	var startKey kv.Key
 	var tableIDs []int64
-	if err := job.DecodeArgs(&tableIDs); err != nil {
-		// arg error, cancel this job.
+	if err := job.DecodeArgs(&tableIDs, &startKey); err != nil {
```
Comment (shenli) [inline ddl/schema.go]: Any kind of compatibility problem with old TiDB cluster?

**Comment by zimulala** [comment]
PTAL @shenli @coocood @disksing

File: ddl/schema.go (hunk 1 above)
Comment (disksing) [inline ddl/schema.go]: I have the same concerns here. We should make sure no this kind of ddl job is running when upgrade.
Or maybe better to add some logic to fallback to table's startKey when it is not set in job.
The rest LGTM.

Comment (zimulala) [inline ddl/schema.go]: If the origin job's arguments are without `startKey`, the decoding out  `startKey` value is nil. And if the value of `startKey` is nil,  the `startKey` will be set to `tablecodec.EncodeTablePrefix(id)`.

Comment (disksing) [inline ddl/schema.go]: I see, LGTM.

File: ddl/reorg.go [hunk 2]
```diff
@@ -169,12 +169,13 @@ func (d *ddl) isReorgRunnable(txn kv.Transaction, flag JobType) error {
 	return nil
 }
 
-// delKeysWithPrefix deletes keys with prefix key in a limited number. If limit < 0, deletes all keys.
-func (d *ddl) delKeysWithPrefix(prefix kv.Key, jobType JobType, job *model.Job, limit int) (int, error) {
+// delKeysWithStartKey deletes keys with start key in a limited number. If limit < 0, deletes all keys.
```
Comment (coocood) [inline ddl/reorg.go]: Add comment about return value.

File: ddl/table.go [hunk 3]
```diff
@@ -137,13 +122,38 @@ func (d *ddl) onDropTable(t *meta.Meta, job *model.Job) error {
...
 	}
 
 	return errors.Trace(err)
 }
 
+// Maximum number of keys to delete for each reorg table job run.
+var reorgTableDeleteLimit = 65536
+
+func (d *ddl) delReorgTable(t *meta.Meta, job *model.Job) error {
+	var prefix kv.Key
```
Comment (coocood) [inline ddl/table.go]: rename it to `startKey`.

File: ddl/table.go [hunk 4]
```diff
@@ -137,13 +122,38 @@ func (d *ddl) onDropTable(t *meta.Meta, job *model.Job) error {
 		job.State = model.JobDone
 		job.SchemaState = model.StateNone
 		addTableHistoryInfo(job, ver, tblInfo)
+		startKey := tablecodec.EncodeTablePrefix(tableID)
+		job.Args = append(job.Args, startKey)
```
Comment (coocood) [inline ddl/table.go]: Can we set `job.Args = []interface{}{startKey}`?

Comment (zimulala) [inline ddl/table.go]: No, we can't. The job has history information in its arguments.

**Comment by coocood** [comment]
LGTM

**Pull request merged**